06.bowling_object/game.rb

```diff
@@ -26,8 +26,8 @@ def record_shot(shot_score)
   end
 
   def apply_bonus(shot_score)
-    @frames.take(9).each do |frame| # ボーナスが適用されるのは9フレーム目まで
-      frame.add_bonus(shot_score)
+    @frames.each do |frame|
+      frame.add_bonus(shot_score) if frame.frame_number < 9
```

Comment:
前回のコメントと同じになりますが、`frame.frame_number < 9`かどうかはframe自身に判定させることができます。

```suggestion
      frame.add_bonus(shot_score)
```

オブジェクトは自分でできることは自分でやらせる、自分でわかることは自分で判断させるのがポイントです。

```ruby
sekito.トイレに行っておいで if sekito.もれそう？
```

ではなく

```ruby
sekito.必要なときにトイレに行きます
```

みたいなイメージ、と言えば伝わりますかね？

大人だったらいちいち外から「もれそう？大丈夫？」って聞かれなくても、「トイレに行くかどうかは自分で決めるわ！（大丈夫かどうかは自分で判断できる）」ってなりますよね。
オブジェクトもそんなふうに「賢く」振る舞うように作るのがポイントです。

また、現時点ですでに `if strike? && @bonuses.size < 2 || spare? && @bonuses.empty?` は自分で判断しているのだから、`if frame.frame_number < 9`だけ外から「どう？」って尋ねるのは実装に一貫性がない、という見方もできます。

Reply:
ちょっと理解出来た気がします！

どうしてここで詰まったかなんですが、僕のイメージとして
```ruby
sekito.トイレに行っておいで if sekito.もれそう？
```
このコードを
```ruby
sekito.必要なときにトイレに行きます
```
こうするためには
```ruby
def 必要なときにトイレに行きます
トイレに行く if もれそう？
end
```

のように定義して、結局もれそうかどうかの判定はしなければいけないので、何か違うのかな？って思ってしまったことが原因だと思っています。（うまく言語化出来てるか怪しいですが）

オブジェクトがメソッドを呼び出した時に条件判定をするよりかは、オブジェクトがメソッドを呼んだらメソッドの中で条件判定して解決したほうが読みやすさとして上だよねという指摘ということであってますでしょうか！

修正したコードではそのように修正したつもりです！

Reply:
> オブジェクトがメソッドを呼び出した時に条件判定をするよりかは、オブジェクトがメソッドを呼んだらメソッドの中で条件判定して解決したほうが読みやすさとして上だよねという指摘ということであってますでしょうか！

はい、そうです。
読みやすさとしてもそうですし、クラスを使う側としてもラク、というメリットがあります。

「add_bonusはframe_numberが8以下のときにだけ呼んでね。9以上のときは呼んじゃダメだぞ、絶対！！」

というクラスより、

「とりあえずadd_bonusを呼んでくれたら、それでOK。あとはいい感じにやっておくんで！」

とした方がクラスを利用する側としては都合がよいです。

今は

- このプログラムを作っているのがsekitoさんだけ
- （業務レベルのコードに比べれば）仕様も単純
- コード量が少ないのでプログラム全体が脳内に収まる

という状態なので、「add_bonusはframe_numberが8以下のときにだけ呼ぶ」という制約を覚えることができますが、

- 自分以外の人がクラスを設計している
- 仕様がめちゃ複雑
- コード量が多すぎてとても脳内に収まらない

という状態になったときは、「○○メソッドはxxxかつxxxかつxxx、またはxxxがxxxのときだけ呼ぶ」みたいなルールがあるとそのルールを理解するのに時間がかかるので開発のスピードが落ちたり、うっかりルールの解釈を間違えてバグを生んだりしてしまいます。

ゆえに、クラス設計はクラスを使う側にとって覚えることが少なく、使い方を間違えにくい（バグを作りにくい）設計にすることが大事になってきます。

https://bootcamp.fjord.jp/reports/95751 でカプセル化しない方が楽なんじゃない？と書いてありましたが、実際はその反対で「このメソッドを使えばそれでOK。クラスの利用者はそれ以外のことは知らなくていいし、こっちとしても教えない、触らせない。バグの原因になるから」というスタンスでクラスを設計した方がみんな幸せになれる、ということですね。

Reply:
クラスはクラスの中で完結させてしまうことでそのクラスの再利用性が高まるんですね。
今まで別にクラスに分けなくても全部トップレベルに書けばいいじゃんのスタンスでいましたが、そう考えるとクラスが少し便利なものに思えました。